File: leetcode/1864.cpp
class Solution {
public:
    int minSwaps(string s) {
        int s0n0 = 0, s0n1 = 0;
        int s1n0 = 0, s1n1 = 0;
        for (int c = 0; c < s.size(); c++) {
            if (c % 2 == 0) {
                if (s[c] != '0')
                    s0n0++;  //对于s0来说，这个位得是0
                else
                    s1n1++;  //对于s1来说，这个位得是1
            } else {
                if (s[c] != '1')
                    s0n1++;  //对于s0来说，这个位得是1
                else
                    s1n0++;  //对于s1来说，这个位得是0
            }
        }
        if (s0n0 != s0n1 && s1n0 != s1n1) return -1;    // s0 s1 都换不了，返回-1
        if (s0n0 == s0n1 && s1n0 != s1n1) return s0n0;  // s0 换得了，返回s0
        if (s0n0 != s0n1 && s1n0 == s1n1) return s1n0;  // s1 换得了，返回s1
        return min(s0n0, s1n0);                         // 两个都换得了，返回交换次数较少的那个。
    }
};

File: leetcode/1864.go
package leetcode

func minSwaps(s string) int {
	s0n0, s0n1, s1n0, s1n1 := 0, 0, 0, 0
	min := func(x, y int) int {
		if x < y {
			return x
		}
		return y
	}
	for i := range s {
		if i%2 == 0 {
			if s[i] == '0' {
				s1n1++
			} else {
				s0n0++
			}
		} else {
			if s[i] == '1' {
				s1n0++
			} else {
				s0n1++
			}
		}
	}
	if s0n0 == s0n1 {
		if s1n0 == s1n1 {
			return min(s0n0, s1n0)
		} else {
			return s0n0
		}
	} else {
		if s1n0 == s1n1 {
			return s1n0
		} else {
			return -1
		}
	}
	return -1
}
